### PR TITLE
PERF: Performance enhancements for the FindArrayStatistics

### PR DESCRIPTION
**Pull request by imikejackson**
Various performance enhancements for Find Array Statistics

### CODE REVIEW

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp [hunk 1]
```diff
@@ -41,29 +41,52 @@ class FindArrayStatisticsByIndexImpl
...
+    const usize numTuples = m_FeatureIds->getNumberOfTuples(); // m_Messenger.getTotalElements();
     const usize numCurrentFeatures = end - start;
 
     std::vector<uint64> length(numCurrentFeatures, 0);
     std::vector<T> min(numCurrentFeatures, std::numeric_limits<T>::max());
     std::vector<T> max(numCurrentFeatures, std::numeric_limits<T>::min());
     std::vector<T> summation(numCurrentFeatures, 0);
+    usize progressCount = 0;
+
+    //  std::cout << "    Starting Min/Max/Summation loop..." << std::endl;
```
Comment (jmarquisbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]:
```suggestion
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp [hunk 2]
```diff
@@ -41,29 +41,52 @@ class FindArrayStatisticsByIndexImpl
...
   void compute(usize start, usize end) const
   {
-    const usize numTuples = m_Source.getNumberOfTuples();
+
+    std::chrono::steady_clock::time_point m_InitialTime = std::chrono::steady_clock::now();
+    auto now = std::chrono::steady_clock::now();
+    usize m_MilliDelay = 1000;
+
+    const std::atomic_bool& shouldCancel = m_Filter->getCancel();
+    const usize numTuples = m_FeatureIds->getNumberOfTuples(); // m_Messenger.getTotalElements();
```
Comment (jmarquisbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]:
```suggestion
    const usize numTuples = m_FeatureIds->getNumberOfTuples(); 
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp [hunk 3]
```diff
@@ -41,29 +41,52 @@ class FindArrayStatisticsByIndexImpl
...
+      }
       if(m_Mask != nullptr && !m_Mask->isTrue(i))
       {
         continue;
       }
       for(usize j = 0; j < numCurrentFeatures; j++)
       {
-        if((*m_FeatureIds)[i] != static_cast<int32>(start + j))
+        int32 featureId = (*m_FeatureIds)[i];
+        if(featureId != static_cast<int32>(start + j))
```
Comment (jmarquisbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]: @imikejackson Was there a reason we split this into 2 lines? This `featureId` isn't actually used again in this scope so it just uses up unnecessary space.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp [hunk 4]
```diff
@@ -235,51 +304,62 @@ class FindArrayStatisticsByIndexImpl
...
+
+    usize numFeatureSources = end - start;
+    // Create the arrays that will collect the values from the arrays. allocate them to the correct size based on the length array
+    std::vector<std::vector<T>> featureSources(numFeatureSources);
+    for(usize featureSourceIndex = 0; featureSourceIndex < numFeatureSources; featureSourceIndex++)
+    {
+      featureSources[featureSourceIndex].reserve(m_LengthArray->operator[](featureSourceIndex + start));
+    }
+
+    // std::cout << start << "    " << end << "  Starting Median Array Loop... " << std::endl;
```
Comment (jmarquisbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]:
```suggestion
```

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp [hunk 5]
```diff
@@ -400,22 +482,42 @@ void FindStatistics(const DataArray<T>& source, const Int32Array* featureIds, co
...
+    // messenger.setProgressIncrement(numFeatures / 100);
+
+    //    ParallelDataAlgorithm dataAlg;
+    //    dataAlg.setRange(0, numFeatures);
+    //    dataAlg.setParallelizationEnabled(true);
+    //    dataAlg.setGrainSize(500);
+    //    dataAlg.execute(FindArrayStatisticsByIndexImpl<T>(inputValues->FindLength, inputValues->FindMin, inputValues->FindMax, inputValues->FindMean, inputValues->FindStdDeviation,
+    //                                                      inputValues->FindSummation, inputValues->FindHistogram, inputValues->MinRange, inputValues->MaxRange, inputValues->UseFullRange,
+    //                                                      inputValues->NumBins, mask, featureIds, source, lengthArray, minArray, maxArray, meanArray, stdDevArray, summationArray, histArray,
+    //                                                      filter));
```
Comment (jmarquisbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]:
```suggestion
```

@imikejackson Is there a reason we are keeping this around or can we get rid of it?

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp [hunk 6]
```diff
@@ -400,22 +482,42 @@ void FindStatistics(const DataArray<T>& source, const Int32Array* featureIds, co
...
+#ifdef COMPLEX_ENABLE_MULTICORE
+    tbb::simple_partitioner simplePartitioner;
+    size_t grainSize = 500;
+    tbb::blocked_range<size_t> tbbRange(0, numFeatures, grainSize);
+    tbb::parallel_for(tbbRange,
+                      FindArrayStatisticsByIndexImpl<T>(inputValues->FindLength, inputValues->FindMin, inputValues->FindMax, inputValues->FindMean, inputValues->FindStdDeviation,
+                                                        inputValues->FindSummation, inputValues->FindHistogram, inputValues->MinRange, inputValues->MaxRange, inputValues->UseFullRange,
+                                                        inputValues->NumBins, mask, featureIds, source, lengthArray, minArray, maxArray, meanArray, stdDevArray, summationArray, histArray, filter),
+                      simplePartitioner);
+#endif
```
Comment (jmarquisbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]: If we are only letting the use calculate this statistic values by index when multicore is enabled, perhaps we should we be sending back a warning or something if they select one of these statistics with this by index option and don't have mutlicore enabled?

**Review by jmarquisbq — CHANGES_REQUESTED** [review]

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp (hunk 6 above)
Comment (imikejackson) [reply src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]: Good catch.. I bet I never went back and figured out the normal array statistics bit when the "by index" is disabled.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp (hunk 3 above)
Comment (imikejackson) [reply src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]: Debugging aid at some point...

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp (hunk 6 above)
Comment (imikejackson) [reply src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindArrayStatistics.cpp]: I addressed it partially by combining some of the stats algorithms like (min and max, summation/mean/stdDev) which help cut down on the time to run those algorithms.

**Pull request merged**